src/Data/Expr.purs

```diff
@@ -2,89 +2,389 @@ module Lynx.Data.Expr where
+  foldr f z x = foldrDefault f z x
+  foldMap f = case _ of
+    DateTime _ -> mempty
+
+instance traversableDateTimeF :: Traversable DateTimeF where
+  sequence = traverse identity
+  traverse f = case _ of
+    DateTime x -> pure (DateTime x)
+
+decodeDateTimeF :: forall a. ExprJSON -> Either String (DateTimeF a)
+decodeDateTimeF { out, param } = case out of
+  "DateTime" -> do
+    x' <- decodeJson param
+    x <- unformat extendedDateTimeFormatInUTC x'
+    pure (DateTime x)
+  _ -> Left "Expected DateTime"
+
+encodeDateTimeF :: forall a. DateTimeF a -> ExprJSON
+encodeDateTimeF = case _ of
+  DateTime x -> base "DateTime" (format extendedDateTimeFormatInUTC x)
+
+isEmptyDateTimeF :: forall a. DateTimeF a -> Boolean
+isEmptyDateTimeF = case _ of
+  DateTime x -> false
+
+printDateTimeF :: forall a. DateTimeF a -> String
+printDateTimeF = case _ of
+  DateTime x -> format extendedDateTimeFormatInUTC x
+
+newtype IntF a
+  = Int Int
+
+derive instance eqIntF :: Eq (IntF a)
+
+derive instance eq1IntF :: Eq1 IntF
+
+derive instance functorIntF :: Functor IntF
+
+derive newtype instance showIntF :: Show (IntF a)
+
+instance foldableIntF :: Foldable IntF where
+  foldl f z x = foldlDefault f z x
+  foldr f z x = foldrDefault f z x
+  foldMap f = case _ of
+    Int _ -> mempty
+
+instance traversableIntF :: Traversable IntF where
+  sequence = traverse identity
+  traverse f = case _ of
+    Int x -> pure (Int x)
+
+decodeIntF :: forall a. ExprJSON -> Either String (IntF a)
+decodeIntF { out, param } = case out of
+  "Int" -> map Int (decodeJson param)
+  _ -> Left "Expected Int"
+
+encodeIntF :: forall a. IntF a -> ExprJSON
+encodeIntF = case _ of
+  Int x -> base "Int" x
+
+isEmptyIntF :: forall a. IntF a -> Boolean
+isEmptyIntF = case _ of
+  Int x -> false
+
+printIntF :: forall a. IntF a -> String
+printIntF = case _ of
+  Int x -> show x
+
+newtype PairF a
+  = Pair { name :: a, value :: a }
+
+derive instance eqPairF :: (Eq a) => Eq (PairF a)
+
+derive instance eq1PairF :: Eq1 PairF
+
+derive instance functorPairF :: Functor PairF
+
+derive newtype instance showPairF :: (Show a) => Show (PairF a)
+
+instance foldablePairF :: Foldable PairF where
+  foldl f z x = foldlDefault f z x
+  foldr f z x = foldrDefault f z x
+  foldMap f = case _ of
+    Pair x -> f x.name <> f x.value
+
+instance traversablePairF :: Traversable PairF where
+  sequence = traverse identity
+  traverse f = case _ of
+    Pair x -> ado
+      name <- f x.name
+      value <- f x.value
+      in Pair { name, value }
+
+decodePairF :: ExprJSON -> Either String (PairF ExprJSON)
+decodePairF { out, param } = case out of
+  "Pair" -> map Pair (decodeJson param)
+  _ -> Left "Expected Pair"
+
+encodePairF :: PairF ExprJSON -> ExprJSON
+encodePairF = case _ of
+  Pair x -> base "Pair" x
+
+isEmptyPairF :: forall a. PairF a -> Boolean
+isEmptyPairF = case _ of
+  Pair x -> false
+
+printPairF :: PairF String -> String
+printPairF = case _ of
+  Pair x -> show x
```

Comment:
Should we just print `name` here?

I'm starting to wonder if we even need `Pair` at all. The motivation for it is to be able to have a key/value pair where the value is something human readable, like "Condé Nast", and the value is machine-readable, i.e. a unique ID like `13`.

I'm not sure that's necessary here, as all these values are just input by the user, they don't correspond to any primary key in any database or anything. That sort of thing will live at the mapping layer, where if we replace `Pair` with `String`, the user can just say company `"Condé Nast"` corresponds to companyId `13` on a campaign.

The main value I see in keeping it, is that we can change the name without changing the value and having to propagate our changes. E.g. changing
```purescript
Pair { name: "Conde Nast", value: "conde-nast" }
```
to
```purescript
Pair { name: "Condé Nast Entertainment", value: "conde-nast" }
```
wouldn't require us to change our mapping function.

We could also have things like
```purescript
options: Array
  [ Pair { name: "Announce", value: 1 }
  , Pair { name: "Launch", value: 1 }
  , Pair { name: "Onsale", value: 2 }
  , Pair { name: "Final", value: 2 }
  ]
```
Where multiple options can have the same value and so map to the same result in our campaign. Though I don't expect that to be a common scenario.

Reply:
More than happy to move to whatever, but as with the `print` changes for arrays, would you be okay with doing these changes in a separate PR?

Reply:
Yup! Mostly thinking out loud, but also looking for feedback on those thoughts. Happy to move the discussion elsewhere though since it's not really relevant to this PR